### DISCUSSION TITLE
Parquet decimal types not read as SQL decimals

### DISCUSSION THREAD

**Discussion by vtslab** [duckdb/duckdb#4687] [Q&A]
I find that decimal types from parquet files are not treated as decimal types in SQL automatically and you have to explicitly cast with:   
`SELECT ID::DECIMAL`
In particular, omitting this makes large joins on decimal columns much slower and much more memory-hungry.  

The questions:
1. Is the need for explicit decimal casts expected behavior?
2. If not, will issues [#4526](https://github.com/duckdb/duckdb/pull/4526)  and [#4540](https://github.com/duckdb/duckdb/issues/4540) also resolve the expected behavior for type selection of decimal columns in parquet files?

**Comment by Alex-Monahan** [comment]
Would you be able to provide a sample file to use for testing?

**Reply by vtslab** [reply]
Fair request, I will look into it this weekend.

**Comment by vtslab** [comment]
Attached is the file [decimals-spark2-legacy.parquet.zip](https://github.com/duckdb/duckdb/files/9594122/decimals-spark2-legacy.parquet.zip).
As the name says it was generated with pyspark==2.4.8 with python decimal.Decimal() instances loaded into an spark-sql dataframe and default options for writing to parquet.
Reading the schema gives:
`SELECT * FROM parquet_schema('decimals-spark2-legacy.parquet'); `   

           name                  type             converted_type           logical_type    
    spark_schema         BOOLEAN                         UTF8                  NaN    
               x         FIXED_LEN_BYTE_ARRAY          DECIMAL                 NaN    

You can see that the 'x' column with decimals does not have a logical_type, while "normal"  parquet files with decimal colums will have something like DecimalType(scale=3, precision=18) there.

**Reply by Mytherin** [reply]
This seems to work as intended. The Parquet file specifies a precision/scale of (38, 18):

```sql
┌─────────────────────────────────────────────────────┬──────────────┬──────────────────────┬─────────────┬─────────────────┬──────────────┬────────────────┬───────┬───────────┬──────────┬──────────────┐
│                      file_name                      │     name     │         type         │ type_length │ repetition_type │ num_children │ converted_type │ scale │ precision │ field_id │ logical_type │
├─────────────────────────────────────────────────────┼──────────────┼──────────────────────┼─────────────┼─────────────────┼──────────────┼────────────────┼───────┼───────────┼──────────┼──────────────┤
│ data/parquet-testing/decimals-spark2-legacy.parquet │ spark_schema │ BOOLEAN              │ 0           │ REQUIRED        │ 1            │ UTF8           │ 0     │ 0         │ 0        │ NULL         │
│ data/parquet-testing/decimals-spark2-legacy.parquet │ x            │ FIXED_LEN_BYTE_ARRAY │ 16          │ OPTIONAL        │ 0            │ DECIMAL        │ 18    │ 38        │ 0        │ NULL         │
└─────────────────────────────────────────────────────┴──────────────┴──────────────────────┴─────────────┴─────────────────┴──────────────┴────────────────┴───────┴───────────┴──────────┴──────────────┘
```

Reading the file, this is exactly what gets generated by DuckDB:

```sql
D select typeof(x) from 'data/parquet-testing/decimals-spark2-legacy.parquet' limit 1;
┌────────────────┐
│   typeof(x)    │
├────────────────┤
│ DECIMAL(38,18) │
└────────────────┘
```

The problem here is perhaps that the precision is very high, which causes larger decimals to be used. When you cast to `DECIMAL` you are effectively casting from `DECIMAL(38,18)` to `DECIMAL(18,3)`. The lower precision allows for faster operations and a lower memory footprint. However, downcasting decimals is lossy, so this optimization cannot be performed by DuckDB itself.

**Reply by vtslab** [reply]
Thanks for the very helpful explanation, which makes me better understand what I am doing. Indeed, the DECIMAL defaults are documented on [https://duckdb.org/docs/sql/data_types/numeric](https://duckdb.org/docs/sql/data_types/numeric). Hopefully, this discussion will also help some other duckdb users encountering parquet files containing join keys with unnecessarily high DECIMAL scale/precision.
